File: problems_list/剑指OfferII091.粉刷房子/剑指OfferII091-粉刷房子.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        int n=costs.size();
        vector<vector<int>> dp(n, vector<int>(3,0));
        for(int j=0; j<3; j++){
            dp[0][j]=costs[0][j];
        }

        for(int i=1; i<n; i++){
            for(int j=0; j<3; j++){
                int prev1=dp[i-1][(j+2)%3];
                int prev2=dp[i-1][(j+1)%3];
                dp[i][j]=min(prev1, prev2)+costs[i][j];
            }
        }
        


        return min(dp[n-1][0], min(dp[n-1][1], dp[n-1][2]));
    }
};

File: problems_list/剑指OfferII091.粉刷房子/剑指OfferII091-粉刷房子.js
/**
 * @param {number[][]} costs
 * @return {number}
 */
var minCost = function(costs) {
    let n=costs.length;
    let dp=new Array(n).fill(0).map(x=>new Array(3).fill(Infinity));
    for(let i=0; i<3; i++)
        dp[0][i]=costs[0][i];
    
    for(let i=1; i<n; i++){
        for(let j=0; j<3; j++){
            for(let k=0; k<3; k++){
                if(j===k) continue;
                dp[i][j]=Math.min(dp[i][j], dp[i-1][k]+costs[i][j]);
            }
        }
    }
    let ans=Infinity;
    for(let i=0; i<3; i++){
        ans=Math.min(ans, dp[n-1][i]);
    }
    return ans;
};